Webapplication should be aware of context root.

guusdk opened the pull request:
The webapplication should be aware of the context root on which it is exposed. Before this commit, the application assumed that its context was the root context of the domain.

This commit adds a configuration property named 'context', that, if set, allows an administrator to define the context root under which the application is made available. When unset, the root context is assumed.

This change is applied to code responsible for:
- determining what a room name is, based on the current URL.
- redirecting the user back to the inital / welcome page.

guusdk commented:
# Don't merge this just yet!

This pull request intends to replace #1202 which became non-rebase-able after all the time it took me to address the issues that were raised by Jitsi devs.

This is a new attempt at preventing the need for Jitsi Meet to be deployed at the root context of a domain, but has an issue on which I seek help: when navigating to the URL on which Jitsi Meet is hosted, the application attempts to enter a room immediately, instead of showing me the welcome page as I'd expect. The room that I'm joining is equal to the context root on which I'm hosting the application, which makes me believe that I missed at least one reference somewhere.

My lack of experience with React works against me here. I've tried replacing the paths in ```react/features/conference/route.js``` and ```features/welcome/route.js```, as those appear to be hard-coded references to the root context. It did not have have the desired effect. As a test, I added my context root, hard-coded, but that resulted in a redirection from http://example.org/mycontextroot/ to http://example.org/mycontextroot/mycontextroot/ instead of showing me a welcome page at http://example.org/mycontextroot/

guusdk commented:
@lyubomir I'd appreciate your help here. For some reason, the page at http://example.org/mycontextroot attempts to make me join a room. I expect the welcome page there. What am I missing?

lyubomir commented:
@guusdk, the string "mycontextroot" is a valid room name so that's how the app interprets it. I'd like us to fix the context root problem that the Web app has been having for a long time. My plan is the following:

The App determines the context root when it mounts i.e. launches, starts as follows: the context root is the pathname without the file name (i.e. the last component that is not followed by a /). For example:
- the context root of "http://example.org/mycontextroot/" will be "/mycontextroot/",
- the context root of "http://example.org/mycontextroot/roomname" will again be "/mycontextroot/".

WDYT?

lyubomir commented:
@guusdk, as we're rewriting the Web app with React, the parsing of window.location to extract the room name happens in the folder react/features. In order to introduce a context root, we'll surely have to modify _getRoomAndDomainFromUrlObject @ https://github.com/jitsi/jitsi-meet/blob/master/react/features/app/functions.native.js#L16.

lyubomir commented on utils.js:
```diff
@@ -12,23 +12,29 @@
 function getRoomName () { // eslint-disable-line no-unused-vars
```
@guusdk, Is the function used at all? I see it has a `eslint-disable-line no-unused-vars` so we should probably check whether it's still used and, it not, we should remove it.

guusdk replied:
I am seeing one reference, in do_external_connect.js

lyubomir commented on conference.js:
```diff
@@ -219,7 +220,16 @@ function maybeRedirectToWelcomePage(options) {
     if (config.enableWelcomePage) {
```
The WelcomePage was rewritten in React i.e. it was moved to the folder react/features so it looks to me like the source code bellow needs to be rewritten to take into account that now the React App is abstracting the relation between window.location and the React Component rendered at that window.location using Route (an abstraction similar to react-router's).

lyubomir commented:
@guusdk, the way I want us to introduce context root is as follows:

1. The feature app (in the directory react/features/app) will determine the context root when the App React Component mounts. It may (turn out to) be necessary to put the determined context root in the Redux state of features/app.

2. We modify _getRoomAndDomainFromUrlObject to make it aware of context root stored in state (as I said above it may be in the Redux state of features/app).

3. In https://github.com/jitsi/jitsi-meet/blob/master/react/features/app/components/App.web.js#L45 we take into account the context root stored in state as well. I expect that will be mostly juggling with the local variable `path`.

lyubomir commented:
It feels like we'll have to clean the domain abstract up first though before we proceed with context root. The current domain abstraction feels insufficient at the moment because we have a Web app host and a Jitsi Meet deployment domain which may be different. As can be seen around `_getRoomAndDomainFromUrlObject`, we parse certain URLs to convert the Web app domain into a Jitsi Meet deployment domain.

guusdk commented:
@lyubomir Thanks for all of your feedback. I really appreciate it. 

I would be very happy to provide answers to the questions that ask, but I think I am unqualified to do that. I am solid Java / back-end developer. With a lot of time and just enough tinkering skills, I was able to make the old javascript / front-end code do what I wanted (at least, a little bit). Although I see how the new React-based code is a massive improvement over the old code, I am now completely lost again, where it comes to applying modifications.

As for the URL-to-room mapping: "everything after the final slash" as you suggest, would work I think, with the exception of resources (css, images) that the page itself needs. The old code appears to work around that by excluding the dot-character from the mapping (the old regex), which makes sure that things like "all.css" are not parsed as a room. I don't know if this is relevant to the React code base at all.

It is a lot to ask, but I would be very grateful for you (instead of me) to implement the changes needed to make Jitsi-meet run on a non-root context. As I am starting to be pressed for time (I've been trying to make this work for weeks now), I would prefer having a "make it work" solution first, which is followed later by a "make it right" solution (in other words: if we can save time by doing the clean up that you mentioned later, I would prefer that).

guusdk commented:
I rebased the PR, and gave it another go.

This is rather coarse, but does this reflect the approach you wanted @lyubomir?

I'm unsure how to add the context to 'state' (I don't get how 'state' is intended to be used). For now, 
* I've passed 'state' to `_getRoomAndDomainFromUrlObject` (where it is ignored and replaced by a hardcoded value).
* I'm interacting with the config to obtain the context value in `App.web.js` `_navigate`

lyubomir commented:
@guusdk, could you please check whether https://github.com/jitsi/jitsi-meet/pull/1295 works for you?

guusdk commented:
@lyubomir you the man! Your changes suffer from a couple of minor issues that I also found with my changes. I'll comment in your PR on those.
